### ISSUE TITLE
Add support for multiple MERGE .. WHEN [ NOT ] MATCHED AND { condition }

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#7291]
In SQL Server (and perhaps other databases), there is an optional `AND { condition }` clause that can be supplied to a different set of `WHEN MATCHED` clauses.

This can perhaps be emulated through a sophisticated transformation as follows:

### Original

```sql
MERGE INTO t
USING s
ON ...
WHEN MATCHED AND p1 THEN UPDATE SET t.x = s.x
WHEN MATCHED AND p2 THEN UPDATE SET t.y = s.y
```

### Emulation

```sql
MERGE INTO t
USING s
ON ...
WHEN MATCHED THEN UPDATE SET 
  t.x = CASE WHEN p1 THEN s.x ELSE t.x END
  t.y = CASE WHEN p2 THEN s.y ELSE t.y END
```

This isn't exactly the same thing, of course. The first query might perform much better than the second, and have less intrusive locking implications, as well as return different update counts (see #10050, for H2)

**Notes:**

- We already support Oracle's `WHERE` clause syntax, which does the same thing as `AND`. Both syntaxes should be equivalent. (Except for the update count in H2: #10050, as well as the behaviour when wrapped in a data change delta table)
- The presence of predicates means that we should be able to have several matched / not matched clauses
- Instead of `UPDATE`, we should be able to issue a `DELETE` clause (also without Oracle's current `WHERE`)

**Support:**

Native support has been observed in:

- [x] Db2
- [x] Derby
- [x] H2
- [x] HSQLDB (only for single `MATCHED` clauses, lacking Feature F313, “Enhanced MERGE statement”)
- [x] SQL Server (only for single `MATCHED` clauses, lacking Feature F313, “Enhanced MERGE statement”)
- [x] Sybase SQL Anywhere
- [x] Vertica

Emulations will be needed in:

- [x] Firebird
- [x] HSQLDB (multi `MATCHED` clauses need to be emulated)
- [x] Oracle
- [x] SQL Server (multi `MATCHED` clauses need to be emulated)
- [x] Teradata

Parser support:

- [x] Parser support

**Comment by lukaseder** [comment]
We already emulate `WHERE` for Derby and SQL Server. Maybe, let's review if there are missing `@Support` values on `WHERE`?

**Comment by lukaseder** [comment]
The parser already supports single `WHEN MATCHED AND` clauses.

**Comment by lukaseder** [comment]
The current state of the support of this feature and its emulation are looking good, in terms of feasibility. However, it becomes clear that supporting both the standard SQL syntax and the Oracle specific `WHERE` clauses adds too much complexity.

We should model only the standard syntax, and if possible, even deprecate the Oracle API, supporting Oracle's syntax only:

- In the parser
- As an emulation of the standard syntax

This means, we need:

- Support for multiple `MERGE .. WHEN NOT MATCHED AND` clauses
- Support for `WHEN MATCHED [ AND ... ] THEN DELETE`

**Comment by lukaseder** [comment]
... Unsupporting Oracle's syntax might means that instead of writing:

```sql
WHEN MATCHED THEN UPDATE SET ... WHERE x DELETE WHERE y
```

They would have to write:

```sql
WHEN MATCHED AND x THEN UPDATE SET ...
WHEN MATCHED AND y THEN DELETE
```

**Comment by katzyn** [comment]
Oracle's syntax can't be easily emulated with the standard syntax.
```SQL
CREATE TABLE S(ID INT PRIMARY KEY, V INT);
INSERT INTO S VALUES (1, 2);
CREATE TABLE T(ID INT PRIMARY KEY, V INT);
INSERT INTO T VALUES (1, 1);
INSERT INTO T VALUES (2, 3);
MERGE INTO T USING S ON (T.ID = S.ID)
WHEN MATCHED THEN
    UPDATE SET T.V = S.V WHERE T.V = 1
    DELETE WHERE T.V IN (2, 3);
SELECT * FROM T;
> ID V
> -- -
> 2  3
```
`DELETE` in Oracle touches only rows that were modified by the `UPDATE` and uses the updated values in its `WHERE` criteria.

If you'll find a way to emulate it properly, please let me know. Support of this weird vendor-specific syntax creates many unwanted complications in H2 too.

**Comment by lukaseder** [comment]
Interesting, I wasn't aware of the fact that the update result was available to the deletion. Our integration tests weren't designed to cover this, and I don't think I've ever needed this chaining. Perhaps that's one of the reasons for the notorious ORA-38104.

To emulate this with standard SQL, I suspect we have to perform some substitutions, then:

```
MERGE INTO T USING S ON (T.ID = S.ID)
WHEN MATCHED THEN
    UPDATE SET T.V = S.V WHERE T.V = 1
    DELETE WHERE T.V IN (2, 3);

-- Is the same as
MERGE INTO T USING S ON (T.ID = S.ID)
WHEN MATCHED 
  AND T.V = 1 
  THEN UPDATE SET T.V = S.V + 1
WHEN MATCHED 
  AND T.V = 1 /* repeat update condition for chaining*/ 
  AND S.V + 1 /* apply substitution T.V = S.V + 1 */ IN (2, 3) 
  THEN DELETE;
```

Seems simple in theory. There are probably edge cases and oversights, as always.

**Comment by lukaseder** [comment]
But anyway. Your feedback is valuable. Even if Oracle syntax can be emulated with standard syntax, it seems to be much more difficult to achieve the inverse, i.e. to emulate standard syntax on Oracle.

**Comment by lukaseder** [comment]
A subtle caveat when emulating Oracle's syntax is that the deletion should be produced first, before the update. Otherwise, things like these might happen:

```sql
CREATE TABLE t (i int NOT NULL PRIMARY KEY, j int NOT NULL UNIQUE);
INSERT INTO t VALUES (1, 1), (2, 2);

MERGE INTO t
USING (SELECT 1)
ON (1 = 1)
WHEN MATCHED THEN UPDATE SET j = 2
WHEN MATCHED AND i = 1 THEN DELETE;
```

The order of matched clauses is relevant, and as such, it can be observed that in Db2, H2, SQL Server, the above leads to a constraint violation exception, whereas this does not:

```sql
MERGE INTO t
USING (SELECT 1)
ON (1 = 1)
WHEN MATCHED AND i = 1 THEN DELETE
WHEN MATCHED THEN UPDATE SET j = 2
```

If the `UPDATE` has no side effects other than updating itself, then we can switch the order since we're substituting (see https://github.com/jOOQ/jOOQ/issues/7291#issuecomment-610378999)

This now behaves in the same way as Oracle's

```sql
MERGE INTO t
USING (SELECT 1 FROM dual)
ON (1 = 1)
WHEN MATCHED THEN UPDATE SET j = 2 DELETE WHERE i = 1;
```

Where `UPDATE` and `DELETE` matching happens in one go, thus avoiding the constraint violation exception

**Comment by lukaseder** [comment]
SQL Server doesn't like this statement either_

```sql
MERGE INTO t
USING (SELECT 1)
ON (1 = 1)
WHEN MATCHED THEN UPDATE SET j = 2
WHEN MATCHED AND i = 1 THEN DELETE;
```

... but for different reasons:

> SQL-Fehler [5324] [S0001]: In a MERGE statement, a 'WHEN MATCHED' clause with a search condition cannot appear after a 'WHEN MATCHED' clause with no search condition.

So, when one `WHEN MATCHED` clause contains a condition, we'll have to add a condition to all of them, e.g. 

```sql
MERGE INTO t
USING (SELECT 1)
ON (1 = 1)
WHEN MATCHED AND 1 = 1 THEN UPDATE SET j = 2
WHEN MATCHED AND i = 1 THEN DELETE;
```

**Comment by lukaseder** [comment]
When emulating a sequence of `WHEN MATCHED` clauses, the question is what the result should be:

```sql
DROP TABLE IF EXISTS t;
CREATE TABLE t (i int NOT NULL PRIMARY KEY, j int);
INSERT INTO t VALUES (1, 1), (2, 2);

MERGE INTO t
USING (SELECT 1)
ON (1 = 1)
WHEN MATCHED THEN UPDATE SET j = 2
WHEN MATCHED AND i = 2 THEN UPDATE SET j = 3;
```

H2 produces (updating the same row several times):

```
I|J|
-|-|
1|2|
2|3|
```

Db2 produces (updating each row only once):

```
I|J|
-|-|
1|2|
2|2|
```

There's also a difference in what is produced as a result when mixing `UPDATE` and `DELETE`:

```sql
MERGE INTO t
USING (SELECT 1)
ON (1 = 1)
WHEN MATCHED AND 1 = 1 THEN UPDATE SET j = 2
WHEN MATCHED AND i = 1 THEN DELETE;
```

H2 produces:

```
I|J|
-|-|
2|2|
```

Db2 produces:

```
I|J|
-|-|
1|2|
2|2|
```

Let's look at the SQL standard to see who's right.

**Comment by lukaseder** [comment]
It seems, Db2 is right as per ISO/IEC 9075-2:2016(E) 14.12 &lt;merge statement> GR 6) a) i) 1) A) I) and II):

> R1 is not a subject row identified by any other &lt;merge when matched clause> that precedes MWMCj  in the &lt;merge operation specification>

**Comment by lukaseder** [comment]
Reported an issue to H2: https://github.com/h2database/h2database/issues/2552. The issue documents a workaround, where in H2, all preceding `WHEN MATCHED` predicates would be negated in subsequent `WHEN MATCHED` clauses:

```sql
MERGE INTO t
USING (SELECT 1)
ON (1 = 1)
WHEN MATCHED 
  /* new predicate */ AND predicate1 THEN UPDATE SET j = 2
WHEN MATCHED 
  /* negation of previous predicates */ AND NOT predicate1 
  /* new predicate */ AND predicate2 THEN UPDATE SET j = 3;
WHEN MATCHED 
  /* negation of previous predicates */ AND NOT predicate1 AND NOT predicate2
  /* new predicate */ AND predicate3 THEN UPDATE SET j = 4;
```

**Comment by lukaseder** [comment]
A full emulation of the F313, “Enhanced MERGE statement” is more tricky than at first thought, in databases that only allow for 1 `WHEN MATCHED THEN UPDATE` and 1 `WHEN MATCHED THEN DELETE` clause, such as e.g. SQL Server.

For example, we have this integration test:

```java
assertEquals(4,
create().mergeInto(TBook())
        .usingDual()
        .on(trueCondition())
        .whenMatchedAnd(TBook_ID().eq(1)).thenDelete()
        .whenMatchedThenUpdate().set(TBook_TITLE(), "ABC")
        .whenMatchedAnd(TBook_ID().eq(3)).thenDelete()
        .whenMatchedAnd(TBook_ID().eq(4)).thenUpdate().set(TBook_TITLE(), "XYZ")
        .execute());

assertEquals(3, fetchCount(TBook()));
assertEquals(asList(2, 3, 4), fetchSorted(TBook_ID()));
assertEquals(asList("ABC", "ABC", "ABC"), fetchSorted(TBook_TITLE()));
```

Which in SQL Server currently produces this wrong emulation:

```sql
merge into [test].[dbo].[t_book]
using (
  select 1 [one]
) t
on 1 = 1
when matched and [test].[dbo].[t_book].[ID] = 4 then update set
  [test].[dbo].[t_book].[TITLE] = case
    when [test].[dbo].[t_book].[ID] = 4 then 'XYZ'
    else [test].[dbo].[t_book].[TITLE]
  end
when matched and (
  [test].[dbo].[t_book].[ID] = 1
  or [test].[dbo].[t_book].[ID] = 3
) then delete;
```

As with H2, we have to negate all the previous predicates in declaration order in each `CASE` expression to make sure we produce the intended behaviour.

**Comment by lukaseder** [comment]
Like H2, HSQLDB also executes all `WHEN MATCHED` clauses, several times per row.

**Comment by lukaseder** [comment]
Additional emulation steps are required in Oracle. See also this comment for details: https://github.com/jOOQ/jOOQ/issues/10052#issuecomment-611525766

**Comment by lukaseder** [comment]
I don't think there's anything missing here that's not covered elsewhere as well.
